test/e2e/wireguard_test.go

```diff
@@ -0,0 +1,121 @@
+// the handshake was established when the wg command line is available.
+func TestWireGuard(t *testing.T) {
+	skipIfNumNodesLessThan(t, 2)
+	skipIfHasWindowsNodes(t)
+	providerIsKind := testOptions.providerName == "kind"
+
+	if !providerIsKind {
+		for _, node := range clusterInfo.nodes {
+			skipIfMissingKernelModule(t, node.name, []string{"wireguard"})
+		}
+	}
+	data, err := setupTest(t)
+	if err != nil {
+		t.Fatalf("Error when setting up test: %v", err)
+	}
+	defer teardownTest(t, data)
+
+	if !providerIsKind {
+		ac := []configChange{
+			{"trafficEncryptionMode", "wireguard", false},
+		}
+		if err := data.mutateAntreaConfigMap(nil, ac, false, true); err != nil {
+			t.Fatalf("Failed to enable WireGuard tunnel: %v", err)
+		}
+		defer func() {
+			ac = []configChange{
+				{"trafficEncryptionMode", "none", false},
+			}
+			if err := data.mutateAntreaConfigMap(nil, ac, false, true); err != nil {
+				t.Fatalf("Failed to disable WireGuard tunnel: %v", err)
+			}
+		}()
+	} else {
+		data.redeployAntrea(t, deployAntreaWireGuardGo)
+		defer data.redeployAntrea(t, deployAntreaDefault)
+	}
+
+	t.Run("testWireGuardTunnelConnectivity", func(t *testing.T) { testWireGuardTunnelConnectivity(t, data) })
+}
+
+func (data *TestData) getWireGuardPeerEndpointsWithHandshake(nodeName string) ([]string, error) {
+	var peerEndpoints []string
+	antreaPodName, err := data.getAntreaPodOnNode(nodeName)
+	if err != nil {
+		return peerEndpoints, err
+	}
+	cmd := []string{"wg"}
+	stdout, stderr, err := data.runCommandFromPod(antreaNamespace, antreaPodName, "wireguard", cmd)
+	if err != nil {
+		return peerEndpoints, fmt.Errorf("error when running 'wg' on '%s': %v - stdout: %s - stderr: %s", nodeName, err, stdout, stderr)
+	}
+	peerConfigs := strings.Split(stdout, "\n\n")
+	if len(peerConfigs) < 1 {
+		return peerEndpoints, fmt.Errorf("invalid 'wg' output on '%s': %v - stdout: %s - stderr: %s", nodeName, err, stdout, stderr)
+	}
+
+	for _, p := range peerConfigs[1:] {
+		lines := strings.Split(p, "\n")
+		if len(lines) < 2 {
+			return peerEndpoints, fmt.Errorf("invalid WireGuard peer config output - %s", p)
+		}
+		peerEndpoint := strings.TrimPrefix(strings.TrimSpace(lines[1]), "endpoint: ")
+		for _, l := range lines {
+			if strings.Contains(l, "latest handshake") {
+				peerEndpoints = append(peerEndpoints, peerEndpoint)
+				break
+			}
+		}
+	}
+	return peerEndpoints, nil
+}
+
+func testWireGuardTunnelConnectivity(t *testing.T, data *TestData) {
+	podInfos, deletePods := createPodsOnDifferentNodes(t, data, "differentnodes")
+	defer deletePods()
+	numPods := 2
+	data.runPingMesh(t, podInfos[:numPods], agnhostContainerName)
+	// wg command is only available in WireGuard sidecar container.
+	if testOptions.providerName == "kind" {
+		nodeName0 := podInfos[0].nodeName
+		nodeName1 := podInfos[1].nodeName
+		endpoints, err := data.getWireGuardPeerEndpointsWithHandshake(nodeName0)
+		assert.NoError(t, err)
```

Comment:
nit: in most cases, it should be `require.NoError` so the test can terminate earlier upon error and the following verification can at least get some valid result, instead of empty values, which may lead to panic in some cases. This applies to the e2e test and all unit tests you add in this patch.

Reply:
ok.